rust/src/shards/gui/layouts/group.rs

```diff
@@ -0,0 +1,185 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright © 2022 Fragcolor Pte. Ltd. */
+
+use super::Group;
+use crate::shard::Shard;
+use crate::shards::gui::util;
+use crate::shards::gui::EGUI_UI_SEQ_TYPE;
+use crate::shards::gui::PARENTS_UI_NAME;
+use crate::types::Context;
+use crate::types::ExposedInfo;
+use crate::types::ExposedTypes;
+use crate::types::InstanceData;
+use crate::types::OptionalString;
+use crate::types::ParamVar;
+use crate::types::Parameters;
+use crate::types::ShardsVar;
+use crate::types::Type;
+use crate::types::Types;
+use crate::types::Var;
+use crate::types::ANY_TYPES;
+use crate::types::SHARDS_OR_NONE_TYPES;
+
+lazy_static! {
+  static ref GROUP_PARAMETERS: Parameters = vec![(
+    cstr!("Contents"),
+    cstr!("The UI contents."),
+    &SHARDS_OR_NONE_TYPES[..],
+  )
+    .into(),];
+}
+
+impl Default for Group {
+  fn default() -> Self {
+    let mut parents = ParamVar::default();
+    parents.set_name(PARENTS_UI_NAME);
+    Self {
+      parents,
+      requiring: Vec::new(),
+      contents: ShardsVar::default(),
+      exposing: Vec::new(),
+    }
+  }
+}
+
+impl Shard for Group {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.Group")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.Group-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.Group"
+  }
+
+  fn help(&mut self) -> OptionalString {
+    OptionalString(shccstr!("Visually groups the contents together."))
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn inputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!(
+      "The value that will be passed to the Contents shards of the group."
+    ))
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn outputHelp(&mut self) -> OptionalString {
+    OptionalString(shccstr!("The output of this shard will be its input."))
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&GROUP_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => self.contents.set_param(value),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.contents.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    let exp_info = ExposedInfo {
+      exposedType: EGUI_UI_SEQ_TYPE,
+      name: self.parents.get_name(),
+      help: cstr!("The parent UI objects.").into(),
+      ..ExposedInfo::default()
+    };
+    self.requiring.push(exp_info);
+
+    Some(&self.requiring)
+  }
+
+  fn exposedVariables(&mut self) -> Option<&ExposedTypes> {
```

Comment:
This seems to repeated some times for different containers, maybe we can have some function for this

Reply:
It's tedious in rust :D but yes you could have a type that implements this body composed inside the actual Shard type and simply do an indirection to it, I think I do that in Physics.

Reply:
I'll see if I can add that to the `util.rs` file.

Reply:
I was able to extract the most logic into utility functions. Also for when exposing the `UI.Parents` variable.